### PR TITLE
format-table: add concept of "vertical" table

### PR DESCRIPTION
**Pull request by poettering**
This adds a concept for "vertical" tables, i.e. where we show information about a single object with field names on the left, and values on the right, like "timedatectl" and so on use it. THis was previously implemented via tables already, but manually. Let's add an official concept for such "vertical" tables. This allows us to give them common styling, and generate useful JSON objects from them.

Basically: if you want to display info about a single object use this "vertical" mode. If you want to show info about multiple objects, use the traditional mode.

This ports over resolvectl and timedatectl. But there's more existing code that could benefit from this, such as networkctl and so on.

### CODE REVIEW

File: src/shared/format-table.c [hunk 1]
```diff
@@ -2761,6 +2817,90 @@ int table_to_json(Table *t, JsonVariant **ret) {
...
+
+                if (i % t->n_columns == 0) {
+                        _cleanup_free_ char *mangled = NULL;
+                        const char *n;
+
+                        n = table_get_json_field_name(t, i / t->n_columns - 1);
+                        if (!n) {
+                                TableData *d;
+
+                                assert_se(d = t->data[i]);
```
Comment (yuwata) [inline src/shared/format-table.c]: `TableData *d = ASSERT_PTR(t->data[i]);`

File: src/shared/format-table.h [hunk 2]
```diff
@@ -12,6 +12,7 @@
 typedef enum TableDataType {
         TABLE_EMPTY,
         TABLE_STRING,
+        TABLE_FIELD, /* used in vertical mode */
```
Comment (yuwata) [inline src/shared/format-table.h:15]: Maybe, it is better to refuse to add cells with `TABLE_FIELD` in horizontal mode?

File: src/shared/format-table.c [hunk 3]
```diff
@@ -216,6 +218,38 @@ Table *table_new_internal(const char *first_header, ...) {
...
+                return NULL;
+
+        t->vertical = true;
+        t->header = false;
+
+        if (table_add_cell(t, &cell, TABLE_STRING, "key") < 0)
+                return NULL;
+
+        if (table_set_uppercase(t, cell, true) < 0)
+                return NULL;
```
Comment (yuwata) [inline src/shared/format-table.c:236]: Why not simply set `"KEY"`?

File: src/shared/format-table.c [hunk 4]
```diff
@@ -216,6 +218,38 @@ Table *table_new_internal(const char *first_header, ...) {
...
+
+        if (table_set_align_percent(t, cell, 100) < 0)
+                return NULL;
+
+        if (table_add_cell(t, &cell, TABLE_STRING, "value") < 0)
+                return NULL;
+
+        if (table_set_uppercase(t, cell, true) < 0)
+                return NULL;
```
Comment (yuwata) [inline src/shared/format-table.c:245]: Also here.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/shared/format-table.h (hunk 2 above)
Comment (poettering) [reply src/shared/format-table.h:15]: so I think sooner or later we might want to allow a "hybrid" concept, where we have the headers up top, but *also* the fields on the left. thus i didn't want to restrict these cells unnecessarily for now. I mean, ultimately they are just string cells with some special formatting.

I think probably should add TABLE_HEADER or so too, and then make the header cells also something that is not just TABLE_STRING.

File: src/shared/format-table.c (hunk 3 above)
Comment (poettering) [reply src/shared/format-table.c:236]: i think there's value in seperating column names from the way they are displayed. i.e, it might make sense to explicitly request traditional JSON output also for vertical tables, and in that case the uppercase field names would be weird...

**Comment by bluca** [comment]
```
2022-11-10T18:17:33.8034521Z stderr:
2022-11-10T18:17:33.8034718Z /* test_issue_9549 */
2022-11-10T18:17:33.8034921Z /* test_multiline */
2022-11-10T18:17:33.8035126Z /* test_strv */
2022-11-10T18:17:33.8035336Z /* test_strv_wrapped */
2022-11-10T18:17:33.8035529Z /* test_json */
2022-11-10T18:17:33.8035722Z /* test_table */
2022-11-10T18:17:33.8035924Z /* test_vertical */
2022-11-10T18:17:33.8036614Z Assertion 'streq(formatted, "     pfft aa: foo\n" "       uuu o: 1.0K\n" "lllllllllllo: jjjjjjjjjjjjjjjjj\n")' failed at src/test/test-format-table.c:555, function void test_vertical(void)(). Aborting.
```

**Review by yuwata — APPROVED** [review]
LGTM.
